File: lr/src/questions/q_0025_reverse_nodes_in_k_group.rs
use crate::utils::ListNode;

pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    let mut head = head;
    let mut next = &mut head;
    loop {
        let mut start = next.take();
        let mut prev = &mut start;
        for _ in 0..k - 1 {
            if prev.is_none() {
                break;
            }
            prev = &mut prev.as_mut().unwrap().next;
        }
        if prev.is_none() {
            *next = start;
            break;
        }
        let mut prev = prev.as_mut().unwrap().next.take();
        while let Some(mut x) = start {
            start = x.next.take();
            x.next = prev;
            prev = Some(x);
        }
        *next = prev;
        for _ in 0..k {
            next = &mut next.as_mut().unwrap().next;
        }
    }
    head
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::list;
    use crate::utils::ListNode;

    #[test]
    fn test_merge_two_lists() {
        assert_eq!(reverse_k_group(list!(1, 2, 3), 2), list!(2, 1, 3));
    }
}

File: lg/src/questions/q_0025_reverse_nodes_in_k_group.go
package questions

import "github.com/hi-rustin/lg/src/utils"

func reverseKGroup(head *utils.ListNode, k int) *utils.ListNode {
	list := make([]int, 0, 2*k)

	current := head
	for current != nil {
		list = append(list, current.Val)
		current = current.Next
	}

	for i := 0; i < len(list); i += k {
		start := i
		end := i + k - 1
		if end >= len(list) {
			break
		}
		for start < end {
			list[start], list[end] = list[end], list[start]
			start++
			end--
		}
	}

	var newHead *utils.ListNode
	next := newHead

	for _, v := range list {
		node := &utils.ListNode{
			Val: v,
		}
		if newHead == nil {
			newHead = node
			next = node
		} else {
			next.Next = node
			next = node
		}
	}

	return newHead
}
